tests/protocol/test_goal_state.py

```diff
@@ -116,26 +118,152 @@ def test_it_should_redact_the_protected_settings_when_saving_to_the_history_dire
+
+            self._assert_goal_state(goal_state, expected_incarnation, channel=GoalStateChannel.WireServer)
+
+            #
+            # Verify update()
+            #
+            timestamp += datetime.timedelta(seconds=15)
+            protocol.mock_wire_data.set_etag('333333', timestamp)
+
+            goal_state.update()
+
+            self._assert_goal_state(goal_state, expected_incarnation, channel=GoalStateChannel.WireServer)
+
+    def test_it_should_use_fast_track_goal_states_from_the_host_ga_plugin(self):
+        with GoalStateTestCase._create_protocol_ws_and_hgap_in_sync() as protocol:
+            protocol.mock_wire_data.set_vm_settings_source(GoalStateSource.FastTrack)
+
+            #
+            # Verify __init__()
+            #
+            expected_etag = '22222'
+            timestamp = datetime.datetime.utcnow() + datetime.timedelta(seconds=15)
+            protocol.mock_wire_data.set_etag(expected_etag, timestamp)
+
+            goal_state = GoalState(protocol.client)
+
+            self._assert_goal_state(goal_state, expected_etag, channel=GoalStateChannel.HostGAPlugin)
+
+            #
+            # Verify update()
+            #
+            expected_etag = '333333'
+            timestamp += datetime.timedelta(seconds=15)
+            protocol.mock_wire_data.set_etag(expected_etag, timestamp)
+
+            goal_state.update()
+
+            self._assert_goal_state(goal_state, expected_etag, channel=GoalStateChannel.HostGAPlugin)
+
+    def test_it_should_use_the_most_recent_goal_state(self):
+        with GoalStateTestCase._create_protocol_ws_and_hgap_in_sync() as protocol:
+            goal_state = GoalState(protocol.client)
+
+            # The most recent goal state is FastTrack
+            timestamp = datetime.datetime.utcnow() + datetime.timedelta(seconds=15)
+            protocol.mock_wire_data.set_vm_settings_source(GoalStateSource.FastTrack)
+            protocol.mock_wire_data.set_etag('222222', timestamp)
+
+            goal_state.update()
+
+            self._assert_goal_state(goal_state, '222222', channel=GoalStateChannel.HostGAPlugin, source=GoalStateSource.FastTrack)
+
+            # The most recent goal state is Fabric
+            timestamp += datetime.timedelta(seconds=15)
+            protocol.mock_wire_data.set_incarnation('222', timestamp)
+
+            goal_state.update()
+
+            self._assert_goal_state(goal_state, '222', channel=GoalStateChannel.WireServer, source=GoalStateSource.Fabric)
+
+            # The most recent goal state is Fabric, but it is coming from the HostGAPlugin (should be ignored)
+            timestamp += datetime.timedelta(seconds=15)
+            protocol.mock_wire_data.set_vm_settings_source(GoalStateSource.Fabric)
+            protocol.mock_wire_data.set_etag('333333', timestamp)
+
+            goal_state.update()
+
+            self._assert_goal_state(goal_state, '222', channel=GoalStateChannel.WireServer, source=GoalStateSource.Fabric)
```

Comment:
I don't know how hard it's, can we verify extension_goal_state it's from HGAP or wirseserver especially when we have both goalstates updated

Reply:
Not sure I understand the question. In this specific assert we are checking that the goal state comes from the WireServer.

Reply:
Is there any where do we have check for extenion_goal_state if it's vm_settings or extension_config?

Reply:
the checks on  channel take care of that. ExtensionsConfig always comes from the WireServer, vmSettings from the HGAP